### DISCUSSION TITLE
Is it possible to exclude variable from an operator, without altering the complexity?

### DISCUSSION THREAD

**Discussion by Zhonglu** [MilesCranmer/PySR#226] [Q&A] [answered]
Hi,

PySR looks promising with some features like "constraints" and "nested_constraints", which Eureqa did not seem to be capable of.

Currently, to exclude the variable from an operator, e.g. "pow", we must play with complexity:
```
    complexity_of_variables = 2,
    complexity_of_constants = 1,
    constraints={
        "pow": (-1, 1),
    },
```
In this way, no variable can appear in the y position of x^y. However, this also makes the constants appears too frequently in the results.

Is there a way to exclude variable from an operator without affecting complexity? I believe it can be useful for implementing domain knowledge.

With kind regards,

Mike

**Comment by MilesCranmer** [comment, marked as answer]
For custom constraints like this, the best thing is to write a function that checks the underlying expressions explicitly. This can be done by modifying the backend as described on this page: https://astroautomata.com/PySR/backend/. You will want to edit the `check_constraints` function to perform this check.
